File: medium/1382/main.c
void DFS (struct TreeNode *root, int *inOrderList, int *top) {
  if (!root) return;
  DFS(root->left, inOrderList, top);
  inOrderList[++(*top)] = root->val;
  DFS(root->right, inOrderList, top);
}
struct TreeNode* CreateBalanceTree (int *inOrderList, int s, int e) {
  if (!(s<=e)) return NULL;
  int mid = (s+e) >> 1;
  struct TreeNode* root = (struct TreeNode*) malloc(sizeof(struct TreeNode));
  root->val = inOrderList[mid];
  root->left = CreateBalanceTree(inOrderList, s, mid-1);
  root->right = CreateBalanceTree(inOrderList, mid+1, e);
  return root;
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* balanceBST(struct TreeNode* root) {
  int *inOrderList = (int*) malloc(10001 * sizeof(int));
  int top = -1;
  DFS(root, inOrderList, &top);
  struct TreeNode* res = CreateBalanceTree(inOrderList, 0, top);
  free(inOrderList);
  return res;
}

File: medium/1382/main.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn balance_bst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {

        let mut InorderList = Vec::new();
        // create inorder list
        Self::CreateInorderList(root.clone(), &mut InorderList);
        let n = InorderList.len();
        // build balance tree
        Self::CreateBalanceTree(&InorderList, 0, (n as i32)-1)
        
    }
    fn CreateInorderList (root: Option<Rc<RefCell<TreeNode>>>, InorderList: &mut Vec<i32>) {
        if let Some(n) = root {
            let node = n.borrow();
            Self::CreateInorderList(node.left.clone(), InorderList);
            InorderList.push(node.val);
            Self::CreateInorderList(node.right.clone(), InorderList);
        }
    }

    fn CreateBalanceTree (InorderList: &Vec<i32>, s:i32, e:i32) -> Option<Rc<RefCell<TreeNode>>> {
        if s > e {
            return None;
        }
        let mid = (s+e) >> 1;

        let mut node = TreeNode::new(InorderList[mid as usize]); 
        node.left = Self::CreateBalanceTree(InorderList, s, mid-1);
        node.right = Self::CreateBalanceTree(InorderList, mid+1, e);

        Some(Rc::new(RefCell::new(node)))
    }
}
